doop-api/metrics.go

```diff
@@ -0,0 +1,201 @@
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-bits/logg"
+
+	"github.com/sapcc/gatekeeper-addons/internal/doop"
+)
+
+// MetricCollector is a prometheus.Collector for all metrics provided by this application.
+type MetricCollector struct {
+	downloader             *Downloader
+	objectIdentityKeys     []string
+	rawViolationsGauge     *prometheus.GaugeVec
+	groupedViolationsGauge *prometheus.GaugeVec
+	auditAgeOldestGauge    *prometheus.GaugeVec
+}
+
+// NewMetricCollector initializes a MetricCollector.
+func NewMetricCollector(downloader *Downloader) *MetricCollector {
+	objectIdentityKeys := strings.Fields(os.Getenv("DOOP_API_OBJECT_IDENTITY_LABELS"))
+
+	//the given key names may not be suitable as label names because of the restricted label name grammar
+	//-> sanitize all non-alphanumeric characters into underscores for the label names
+	objectIdentityLabels := make([]string, len(objectIdentityKeys))
+	rx := regexp.MustCompile(`[^a-zA-Z0-9]`)
+	for idx, key := range objectIdentityKeys {
+		objectIdentityLabels[idx] = rx.ReplaceAllString(key, "_")
+	}
+
+	return &MetricCollector{
+		downloader:         downloader,
+		objectIdentityKeys: objectIdentityKeys,
+		rawViolationsGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "doop_raw_violations",
+				Help: "Number of raw violations, grouped by constraint, source cluster and selected object identity labels.",
+			},
+			append([]string{"cluster", "template_kind", "constraint_name"}, objectIdentityLabels...),
+		),
+		groupedViolationsGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "doop_grouped_violations",
+				Help: "Number of violation groups, grouped by constraint, source cluster and selected object identity labels.",
+			},
+			append([]string{"template_kind", "constraint_name"}, objectIdentityLabels...),
+		),
+		auditAgeOldestGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "doop_oldest_audit_age_seconds",
+				Help: "Data age for each source cluster.",
+			},
+			[]string{"cluster"},
+		),
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (mc *MetricCollector) Describe(ch chan<- *prometheus.Desc) {
+	mc.rawViolationsGauge.Describe(ch)
+	mc.groupedViolationsGauge.Describe(ch)
+	mc.auditAgeOldestGauge.Describe(ch)
+}
+
+// Collect implements the prometheus.Collector interface.
+func (mc *MetricCollector) Collect(ch chan<- prometheus.Metric) {
+	descCh := make(chan *prometheus.Desc, 1)
+	mc.rawViolationsGauge.Describe(descCh)
+	rawViolationsDesc := <-descCh
+	mc.groupedViolationsGauge.Describe(descCh)
+	groupedViolationsDesc := <-descCh
+	mc.auditAgeOldestGauge.Describe(descCh)
+	auditAgeOldestDesc := <-descCh
+
+	//using the individual reports, we can immediately calculate the audit age
+	reports, err := mc.downloader.GetReports()
+	if err != nil {
+		logg.Error("could not download reports for metric computation: %s", err.Error())
```

Comment:
Shouldn't we stop here if an error occurs?

Reply:
I don't think it's necessary to take the entire doop-api down over this. We will see failed metric scraping through absence alerts.